java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java

```diff
@@ -905,43 +906,87 @@ public ResponseEntity<String> confirmCode(String code, PremiumUserDevicesReposit
+        pu.token = null;
+        pu.tokenTime = null;
+        usersRepository.saveAndFlush(pu);
+    }
+    
     
-    public ResponseEntity<String> changeEmail(MapApiController.UserPasswordPost us, PremiumUserDevicesRepository.PremiumUserDevice dev, HttpServletRequest request) throws ServletException {
-        PremiumUsersRepository.PremiumUser pu = usersRepository.findById(dev.userid);
-        if (pu == null) {
+    public ResponseEntity<String> changeEmail(String username, String token, PremiumUserDevicesRepository.PremiumUserDevice dev, HttpServletRequest request) throws ServletException {
+        // validate new email
+        PremiumUsersRepository.PremiumUser tempUser = usersRepository.findByEmail(username);
+        if (tempUser == null) {
+            return ResponseEntity.badRequest().body("Something went wrong with your new email");
+        }
+        if (tempUser.orderid != null) {
+            String errorMsg = userSubService.checkOrderIdPremium(tempUser.orderid);
+            if (errorMsg != null) {
+                return ResponseEntity.badRequest().body("You can't change email, because you have subscription on new email");
+            }
+        } else {
+            List<PremiumUserFilesRepository.UserFileNoData> allFiles = filesRepository.listFilesByUserid(tempUser.id, null, null);
+            if (!allFiles.isEmpty()) {
+                return ResponseEntity.badRequest().body("You can't change email, because you have files in account on new email");
+            }
+        }
+        List<PremiumUserDevicesRepository.PremiumUserDevice> devices = devicesRepository.findByUserid(tempUser.id);
+        if (!devices.isEmpty()) {
+            return ResponseEntity.badRequest().body("You can't change email, because you have devices in account on new email");
+        }
+        // validate token
+        ResponseEntity<String> response = confirmCode(username, token);
+        if (!response.getStatusCode().is2xxSuccessful()) {
+            return response;
+        }
+        // validate current user
+        PremiumUsersRepository.PremiumUser currentUser = usersRepository.findById(dev.userid);
+        if (currentUser == null) {
             return ResponseEntity.badRequest().body("User is not registered");
         }
-        try {
-            boolean tokenExpired = System.currentTimeMillis() - pu.tokenTime.getTime() > TimeUnit.MILLISECONDS.convert(24, TimeUnit.HOURS);
-            if (pu.token.equals(us.token) && !tokenExpired) {
-                pu.email = us.username;
-                usersRepository.saveAndFlush(pu);
-                request.logout();
-                return ok();
+        // change email
+        usersRepository.delete(tempUser);
+        currentUser.email = username;
+        usersRepository.saveAndFlush(currentUser);
+        request.logout();
+        
+        return ok();
+    }
+    
+    @Scheduled(fixedDelay = 1000 * 60 * 60 * 24)
```

Comment:
Not sure this needs to be done so far we can do it manually with review sometimes